File: LeetCode/0108.Convert-Sorted-Array-To-Binary-Search-Tree/Convert-Sorted-Array-To-Binary-Search-Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        length = len(nums)
        if length == 0:
            return None

        mid_pos = length // 2
        node = TreeNode(nums[mid_pos])
        if mid_pos == 0:
            return node

        node.left = self.sortedArrayToBST(nums[:mid_pos])
        node.right = self.sortedArrayToBST(nums[mid_pos + 1:])

        return node

File: LeetCode/0108.Convert-Sorted-Array-To-Binary-Search-Tree/Convert-Sorted-Array-To-Binary-Search-Tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 *           7
 *      3         11
 *    1   5     9     13
 *   0 2 4 6  8  10 12 14
 */
 
struct TreeNode* sortedArrayToBST(int* nums, int numsSize) {
    //return buildTree(nums, 0, numsSize);
    if (numsSize==0) return NULL;
    int mid = numsSize/2;
    struct TreeNode* root = malloc(sizeof(struct TreeNode));
    root->val = nums[mid]; 
    root->left = sortedArrayToBST(nums, mid);
    root->right = sortedArrayToBST(nums+mid+1, numsSize-mid-1);
    return root;
}
